cpp/src/arrow/compute/kernels/scalar_replace.cc

```diff
@@ -0,0 +1,309 @@
+                           ctx->AllocateBitmap(data.length));
+
+    if (mask.MayHaveNulls()) {
+      ::arrow::internal::BitmapAnd(mask.buffers[0]->data(), mask.offset,
+                                   mask.buffers[1]->data(), mask.offset, mask.length,
+                                   output->offset, out_nulls->mutable_data());
+      ::arrow::internal::BitmapOr(data.buffers[0]->data(), data.offset, out_nulls->data(),
+                                  output->offset, data.length, output->offset,
+                                  out_nulls->mutable_data());
+    } else {
+      ::arrow::internal::BitmapOr(data.buffers[0]->data(), data.offset,
+                                  mask.buffers[1]->data(), mask.offset, mask.length,
+                                  output->offset, out_nulls->mutable_data());
+    }
+
+    if (::arrow::internal::CountSetBits(out_nulls->data(), output->offset, data.length) <
+        data.length)
+      output->buffers[0] = out_nulls;
+  }
+}
+
+template <typename Type, typename Enable = void>
+struct ReplaceFunctor {};
+
+// Numeric inputs
+
+template <typename Type>
+struct ReplaceFunctor<Type, enable_if_t<is_number_type<Type>::value>> {
+  using T = typename TypeTraits<Type>::CType;
+
+  static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const ArrayData& data = *batch[0].array();
+    const ArrayData& mask = *batch[1].array();
+    const Scalar& replacement = *batch[2].scalar();
+    ArrayData* output = out->mutable_array();
+
+    // Ensure the kernel is configured properly to have no validity bitmap /
+    // null count 0 unless we explicitly propagate it below.
+    DCHECK(output->buffers[0] == nullptr);
+
+    if (replacement.is_valid) {
+      handle_nulls(ctx, data, mask, output);
+
+      KERNEL_ASSIGN_OR_RAISE(std::shared_ptr<Buffer> out_buf, ctx,
+                             ctx->Allocate(data.length * sizeof(T)));
+      T value = UnboxScalar<Type>::Unbox(replacement);
+      const uint8_t* to_replace = mask.buffers[1]->data();
+      const T* in_values = data.GetValues<T>(1);
+      T* out_values = reinterpret_cast<T*>(out_buf->mutable_data());
+      int64_t offset = data.offset;
+      BitBlockCounter bit_counter(to_replace, data.offset, data.length);
+      while (offset < data.offset + data.length) {
+        BitBlockCount block = bit_counter.NextWord();
+        if (block.NoneSet()) {
+          std::memcpy(out_values, in_values, block.length * sizeof(T));
+        } else if (block.AllSet()) {
+          std::fill(out_values, out_values + block.length, value);
+        } else {
+          for (int64_t i = 0; i < block.length; ++i) {
+            out_values[i] =
+                BitUtil::GetBit(to_replace, offset + i) ? value : in_values[i];
+          }
+        }
+        offset += block.length;
+        out_values += block.length;
+        in_values += block.length;
+      }
+      output->buffers[1] = out_buf;
```

Comment:
minor comment, you could use `std::move` here. (applies to several places, where buffer shared_ptr's are assigned)